In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option('display.max_columns', 100)

## Extract information from columns descriptions using the text file `dataset dictionary.txt`

In [2]:
content = []
with open('dataset dictionary.txt', 'r') as f:
    for line in f.readlines():
        content.append(line)

In [3]:
content = [i.strip('\n').split(': ') for i in content]

In [4]:
content = {i[0]: i[1] for i in content} 

In [5]:
var_descrption_df = pd.DataFrame({'variable': content.keys(), 'description': content.values()}, index = np.arange(len(content)))

In [6]:
var_descrption_df

variable                                        description
0      photoUrl                     The URL of the player's photo.
1      LongName                       The full name of the player.
2     playerUrl        The URL of the player's page on sofifa.com.
3   Nationality                     The nationality of the player.
4     Positions                 The positions the player can play.
..          ...                                                ...
72          PAS  This refers to the player's passing ability. I...
73          DRI  This refers to the player's dribbling ability....
74          DEF  This refers to the player's defensive ability....
75          PHY  This refers to the player's physicality or str...
76         Hits  This refers to the number of times the player'...

[77 rows x 2 columns]

In [7]:
var_descrption_df.to_csv('variable_descriptions.csv', index=False)

# Start of cleaning

In [8]:
df = pd.read_csv('fifa21 raw data v2.csv', low_memory=False)

In [9]:
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

Wow!! 77 columns. Let's see them more in depth:

In [10]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

## Missing values study

Let's study the missing values and their counts  
It's not the prettiest method but I can't remember now how to do it in a more elegant way...

In [11]:
s = pd.Series([df[i].isnull().sum() for i in df.columns], index=df.columns).sort_values(ascending=False)
s.name = 'missing values count'
s

Loan Date End       17966
Hits                 2595
GK Diving               0
Sliding Tackle          0
Standing Tackle         0
                    ...  
Heading Accuracy        0
Finishing               0
Crossing                0
Attacking               0
Reactions               0
Name: missing values count, Length: 77, dtype: int64

Study of the dtypes:

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

## Study of object variables

See if any object (mainly string) variable should be numeric

In [13]:
df.select_dtypes(object).columns

Index(['Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Club',
       'Contract', 'Positions', 'Height', 'Weight', 'Preferred Foot',
       'Best Position', 'Joined', 'Loan Date End', 'Value', 'Wage',
       'Release Clause', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'Hits'],
      dtype='object')

In [14]:
# just take 5 rows to check the values and see if they effectively must be integers or some need a type conversion
string_vars = df.select_dtypes(object).sample(5)
string_vars.T

7269   \
Name                                                 M. te Wierik   
LongName                                           Mike te Wierik   
photoUrl         https://cdn.sofifa.com/players/200/425/21_60.png   
playerUrl       http://sofifa.com/player/200425/mike-te-wierik...   
Nationality                                           Netherlands   
Club                                         \n\n\n\nDerby County   
Contract                                              2020 ~ 2023   
Positions                                                      CB   
Height                                                      191cm   
Weight                                                       73kg   
Preferred Foot                                              Right   
Best Position                                                  CB   
Joined                                                Jul 1, 2020   
Loan Date End                                                 NaN   
Value                                                       €1.1M   
Wage                                                         €11K   
Release Clause                                              €1.7M   
W/F                                                           3 ★   
SM                                                             2★   
A/W                                                        Medium   
D/W                                                        Medium   
IR                                                            1 ★   
Hits                                                            7   

                                                            9454   \
Name                                                    L. Pirard   
LongName                                             Lucas Pirard   
photoUrl         https://cdn.sofifa.com/players/223/957/21_60.png   
playerUrl       http://sofifa.com/player/223957/lucas-pirard/2...   
Nationality                                               Belgium   
Club                                     \n\n\n\nWaasland-Beveren   
Contract                                              2019 ~ 2022   
Positions                                                      GK   
Height                                                      191cm   
Weight                                                       91kg   
Preferred Foot                                              Right   
Best Position                                                  GK   
Joined                                               Jul 11, 2019   
Loan Date End                                                 NaN   
Value                                                       €875K   
Wage                                                          €2K   
Release Clause                                              €915K   
W/F                                                           2 ★   
SM                                                             1★   
A/W                                                        Medium   
D/W                                                        Medium   
IR                                                            1 ★   
Hits                                                            2   

                                                            7701   \
Name                                                  M. Pospíšil   
LongName                                          Martin Pospíšil   
photoUrl         https://cdn.sofifa.com/players/203/603/21_60.png   
playerUrl       http://sofifa.com/player/203603/martin-pospisi...   
Nationality                                        Czech Republic   
Club                                \n\n\n\nJagiellonia Białystok   
Contract                                              2017 ~ 2021   
Positions                                                CDM, CAM   
Height                                                      178cm   
Weight                                                  

## Proposed actions

Some conclusions after observing the data from some `object` type columns:  
1. Rename columns so it's easier to access them using dot notation (df.wage)
2. `Height` could be transformed to feet and converted to float
3. `Weight` can be converted to integer after removing lbs
4. `Foot` definitely must be converted to categorical
5. `Value` and `Release Clause` can be converted to float and remove euro sign (or maybe there's some dollar sign)
6. `W/F`, `SM` and `IR` must be converted to integer (or categorical and then ordered category)
7. `Joined` must be datetime dtype
8. `Hits` can be cleaned and transformed into integer (later converted to ordered categorical)
9. Set `a/w` and `d/w` to categorical and sort their category values
10. Study `contract` column
11. Clean `club` values
12. Reestructuration and/or removal of columns after all processing

### 1 Rename columns

In [15]:
columns_backup = df.columns.to_list()
df.columns = [i.replace(' & ', '_').replace(' ', '_').lower() for i in df.columns]

In [16]:
df.columns

Index(['id', 'name', 'longname', 'photourl', 'playerurl', 'nationality', 'age',
       '↓ova', 'pot', 'club', 'contract', 'positions', 'height', 'weight',
       'preferred_foot', 'bov', 'best_position', 'joined', 'loan_date_end',
       'value', 'wage', 'release_clause', 'attacking', 'crossing', 'finishing',
       'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling',
       'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement',
       'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance',
       'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots',
       'mentality', 'aggression', 'interceptions', 'positioning', 'vision',
       'penalties', 'composure', 'defending', 'marking', 'standing_tackle',
       'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling',
       'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats',
       'base_stats', 'w/f', 'sm', 'a/w', 'd/w', 'ir', 'pac', 'sho', 'pas',
       'dri', 'def', 

### 2. `Height` could be transformed to feet and converted to float

In [17]:
height = df.height
height.sample(10)

6493     173cm
16328    167cm
11762    179cm
16635    179cm
5845     184cm
14884    175cm
3602     181cm
5908     184cm
17808    187cm
2279     170cm
Name: height, dtype: object

All heights seem to be in cm. Let's see if there are some feet or inches

In [18]:
# number of row with cm in the value
height.str.contains('ft').sum()

0

In [19]:
# number of row with cm in the value
height.str.contains('inch').sum()

0

In [20]:
# number of row with cm in the value
height.str.contains('feet').sum()

0

All of the heights (which are not null) are in cm?

In [21]:
df.height[df.height.notnull()].str.endswith('cm').all()

False

Let's check what other units there are.

In [22]:
notnull_height = df.height.dropna()

In [23]:
notnull_height.str[3:].unique()

array(['cm', '"', '1"', '0"'], dtype=object)

The results show that there are some values in inches ( " )

In [24]:
notnull_height[notnull_height.str.endswith('"')]

793     6'2"
847     6'3"
848     6'5"
860    5'11"
861     6'4"
862     6'1"
863     6'0"
864     6'1"
865    5'11"
866     6'2"
867     6'0"
868     6'3"
869     6'0"
871    5'10"
872     5'9"
873    5'11"
874     5'6"
875     6'1"
876     6'0"
877     6'4"
878     5'7"
879     6'0"
880    5'11"
881    5'11"
883     6'0"
884    5'10"
885     6'3"
886     6'0"
887     5'4"
888    5'11"
889     5'9"
890     5'7"
891     6'0"
892    5'10"
893     6'2"
894     6'2"
895     6'3"
896     5'7"
897    5'10"
898     5'9"
Name: height, dtype: object

Definitely `height` needs some cleaning and uniform units for height. I will create `height_m` and `height_feet` columns.

In [25]:
df['height_m'] = df.height
df['height_feet'] = df.height

1 metre = 0.3048 feet
1 inch = 1/12 feet

In [26]:
def to_metres(value):
    # if unit is cm
    if value.endswith('cm'):
        value = int(value[:-2])
        # conversion from cm to metre
        value = value / 100
        
    # if value is in feet and inches
    else:
        value = value.strip('"').split("'")
        # conversion from feet and inches to just feet
        value = float(value[0]) + float(value[1])/12
        # conversion from feet to metre
        value *= 0.3048
        
    return value

In [27]:
def to_feet(value):
    # if unit is cm
    if value.endswith('cm'):
        value = int(value[:-2])
        # conversion from cm to feet
        value = value / 30.48
        
    # if value is in feet and inches
    else:
        value = value.strip('"').split("'")
        # conversion to feet
        value = float(value[0]) + float(value[1])/12
        
    return value

In [28]:
df.height_m = df.height_m.apply(to_metres)

In [29]:
df.height_feet = df.height_feet.apply(to_feet)

At the end, I will remove the unnecessary columns. By now, I will keep all the height columns.

### 3. `Weight` can be converted to integer after removing `kg`

In [30]:
weight = df.weight
weight.sample(5)

16875    71kg
6736     81kg
14980    72kg
15037    68kg
8066     69kg
Name: weight, dtype: object

#### Error: `ValueError: invalid literal for int() with base 10: '183l'`
Uncomment the cell below to see the whole error output.

In [31]:
#weight = weight.apply(lambda x : int(x[:-2]) )

It seems as if some weights are in lb and others in Kg

In [32]:
weight.str[-3:].unique()

array(['2kg', '3kg', '7kg', '0kg', '8kg', '1kg', '5kg', '9kg', '4kg',
       '6kg', 'lbs'], dtype=object)

In [33]:
def convert_weight(value):
    if value.endswith('kg'):
        return float(value[:-2])  # lbs to kg will throw float values
    else:
        value = value[:-3]
        value = float(value) / 2.204623  # 1 Kg = 2,204623 lb
        return value

In [34]:
weight = weight.apply(convert_weight)

In [35]:
weight

0        72.0
1        83.0
2        87.0
3        70.0
4        68.0
         ... 
18974    66.0
18975    65.0
18976    74.0
18977    69.0
18978    75.0
Name: weight, Length: 18979, dtype: float64

In [36]:
df.weight = weight

### 4. `Preferred Foot` definitely must be converted to categorical

In [37]:
df.preferred_foot

0         Left
1        Right
2        Right
3        Right
4        Right
         ...  
18974    Right
18975    Right
18976    Right
18977    Right
18978     Left
Name: preferred_foot, Length: 18979, dtype: object

In [38]:
foot =df.preferred_foot
foot.unique()

array(['Left', 'Right'], dtype=object)

In [39]:
df.preferred_foot = foot.astype('category')

In [40]:
df.dtypes['preferred_foot']

CategoricalDtype(categories=['Left', 'Right'], ordered=False)

### 5. `value`, `wage` and `release_clause` can be converted to float and remove euro sign (or maybe there's some dollar sign)

In [41]:
value = df.value
wage = df.wage
release_clause = df.release_clause

In [42]:
wage.sample(5)

3787     €35K
6518      €5K
3750     €13K
13105    €700
10695     €2K
Name: wage, dtype: object

In [43]:
value.sample(5)

16886    €350K
10390    €625K
8117     €900K
7835     €1.1M
8617     €1.1M
Name: value, dtype: object

In [44]:
release_clause.sample(5)

13801    €723K
9271     €831K
11629    €984K
15576    €642K
15143    €189K
Name: release_clause, dtype: object

#### 5.1 Check if all values are in euros or there are in dollars

In [45]:
value.str[0].unique()

array(['€'], dtype=object)

In [46]:
wage.str[0].unique()

array(['€'], dtype=object)

In [47]:
release_clause.str[0].unique()

array(['€'], dtype=object)

#### 5.2 Remove euro sign

In [48]:
value = value.str[1:]
wage = wage.str[1:]
release_clause = release_clause.str[1:]

#### 5.3 Convert to million euros using a custom function

In [49]:
def to_millions(i):
    if i[-1] == 'K':
        return float(i[:-1]) / 1000
    else:
        return float(i[:-1])

#### Error: `ValueError: could not convert string to float: ''`
Uncomment the cell below to see the whole error output.

In [50]:
#value.apply(to_millions)

It looks as if there are some empty values. Count the nan for every variable

In [51]:
v = value.str[:-1]
(v == '').sum()

248

In [52]:
w = wage.str[:-1]
(w == '').sum()

237

In [53]:
r = release_clause.str[:-1]
(r == '').sum()

1261

Last count of nan values was really high. What is the percentage of nan in the variable?

In [54]:
nan_count = (r == '').sum()
print(f'The percentage of nan values in release_clause variable is: {round(nan_count / len(r) * 100, 2)}')

The percentage of nan values in release_clause variable is: 6.64


So the count was not so high in the end.

There are empty values: ''. I will use a slightly modified version of function `to_millions` that replaces the '' values for '0M'.

In [55]:
def to_millions(i):
    if i[:-1] == '':
        return np.nan
    elif i[-1] == 'K':
        return float(i[:-1]) / 1000
    else:
        return float(i[:-1])

In [56]:
value = value.apply(to_millions)
wage = wage.apply(to_millions)
release_clause = release_clause.apply(to_millions)

#### 5.4 Impute nan values by the mean value

Remember that there were empty values, which are now nan values:

In [57]:
value.isna().sum()

248

I will impute nan values by the mean of every variable

In [58]:
value_mean = value[value.notnull()].mean()
value_mean

2.9029965832043136

In [59]:
wage_mean = wage[wage.notnull()].mean()
wage_mean

12.132925301461956

In [60]:
release_clause_mean = release_clause[release_clause.notnull()].mean()
release_clause_mean

4.24499678293261

In [61]:
value = value.fillna(value_mean)
wage = wage.fillna(wage_mean)
release_clause = release_clause.fillna(release_clause_mean)

#### 5.5 Save the clean variables to the dataframe and rename column to `euros_M`

In [62]:
df.value = value
df.wage = wage
df.release_clause = release_clause

In [63]:
df.rename(columns = {'value' : 'value_euros_M',
                    'wage': 'wage_euros_M',
                    'release_clause': 'release_clause_euros_M'},
          inplace='True')

### 6. `W/F`, `SM` and `IR` must be converted to integer (or categorical and then ordered category)

In [64]:
wf = df['w/f']
sm = df.sm
ir = df.ir

Inspect wf, sm and ir values

In [65]:
wf.unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

In [66]:
sm.unique()

array(['4★', '5★', '1★', '2★', '3★'], dtype=object)

In [67]:
ir.unique()

array(['5 ★', '3 ★', '4 ★', '2 ★', '1 ★'], dtype=object)

Custom function to process data:

In [68]:
def remove_star(i):
    return int(i[0])

In [69]:
wf.apply(remove_star)

0        4
1        4
2        3
3        5
4        5
        ..
18974    2
18975    2
18976    2
18977    3
18978    3
Name: w/f, Length: 18979, dtype: int64

In [70]:
wf = wf.apply(remove_star)
sm = sm.apply(remove_star)
ir = ir.apply(remove_star)

Now lets convert to category dtype

In [71]:
wf = wf.astype('category')
sm = sm.astype('category')
ir = ir.astype('category')

Check if categories are ordered:

#### Error: `TypeError: Unordered Categoricals can only compare equality or not`
Uncomment the cell below to see the whole error output.

In [72]:
#ir[ir >=3]

Set ordered categories using `.cat.set_categories([ordered values], ordered=True)`

In [73]:
wf = wf.cat.set_categories([1,2,3,4,5], ordered=True)
sm = sm.cat.set_categories([1,2,3,4,5], ordered=True)
ir = ir.cat.set_categories([1,2,3,4,5], ordered=True)

Check if categories are ordered (again):

In [74]:
ir[ir >=3]

0        5
1        5
2        3
3        4
4        5
        ..
4988     3
5792     3
6286     3
8552     3
10101    3
Name: ir, Length: 332, dtype: category
Categories (5, int64): [1 < 2 < 3 < 4 < 5]

Finally, assign to the original dataframe

In [75]:
df['w/f'] = wf
df.sm = sm
df.ir = ir

### 7. `Joined` must be datetime dtype

In [76]:
joined = df.joined

In [77]:
pd.to_datetime(joined)

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: joined, Length: 18979, dtype: datetime64[ns]

In [78]:
pd.to_datetime(joined).isna().sum()

0

Everything went right, so let's assign

In [79]:
df.joined = pd.to_datetime(joined)

### 8. `Hits` can be cleaned and transformed into integer

In [80]:
hits = df.hits
hits

0        771
1        562
2        150
3        207
4        595
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: hits, Length: 18979, dtype: object

We see there are some 'NaN' values. Let's make a count of nulls

In [81]:
df.hits.isnull().sum()

2595

Let's try a dtype conversion and see what happens...

#### Error: `ValueError: invalid literal for int() with base 10: '1.6K'`
Uncomment the cell below to see the whole error output.

In [82]:
#df.hits.astype(int)

Looks like some of them have letters (K for thousands). They need a deeper transformation.

In [83]:
hits.str.contains('K').sum()

28

In [84]:
hits.str.lstrip()

0        771
1        562
2        150
3        207
4        595
        ... 
18974    NaN
18975    NaN
18976    NaN
18977    NaN
18978    NaN
Name: hits, Length: 18979, dtype: object

In [85]:
def clean_hits(i):
    i = i.lstrip()
    if i.endswith('K'):
        i = i[:-1]
        i = float(i) * 1000
        i = int(i)
    else:
        i = int(i)
    return i

#### Error: `AttributeError: 'float' object has no attribute 'lstrip'`
Uncomment the cell below to see the whole error output.

In [86]:
#hits = hits.apply(clean_hits)

There's an error when processing because NaN values are of type `float`. I will convert them to 0 and then process.

In [87]:
hits = hits.fillna('0')

In [88]:
hits = hits.apply(clean_hits)

Now convert the 0 values to NaN again, so they can be handled later.

In [89]:
hits[hits == 0] = np.nan

In [90]:
hits.isna().sum()

2595

Everything is ok now. Assign to dataframe column:

In [91]:
df.hits = hits

### After a first cleaning stage, let's see how things are now

In [92]:
# just take 5 rows to check the values and see if they effectively must be integers or some need a type conversion
string_vars = df.select_dtypes(object).sample(5)
string_vars.T

18160  \
name                                                M. Al Burayh   
longname                                      Mohammad Al Burayh   
photourl        https://cdn.sofifa.com/players/210/750/21_60.png   
playerurl      http://sofifa.com/player/210750/mohammad-al-bu...   
nationality                                         Saudi Arabia   
club                                            \n\n\n\nAl Fateh   
contract                                             2019 ~ 2022   
positions                                                     GK   
height                                                     181cm   
best_position                                                 GK   
loan_date_end                                                NaN   
a/w                                                       Medium   
d/w                                                       Medium   

                                                           2482   \
name                                                    F. Coman   
longname                                          Florinel Coman   
photourl        https://cdn.sofifa.com/players/238/274/21_60.png   
playerurl      http://sofifa.com/player/238274/florinel-coman...   
nationality                                              Romania   
club                                       \n\n\n\nFCSB (Steaua)   
contract                                             2017 ~ 2022   
positions                                                     LW   
height                                                     182cm   
best_position                                                CAM   
loan_date_end                                                NaN   
a/w                                                         High   
d/w                                                          Low   

                                                           1897   \
name                                                  K. Lasagna   
longname                                           Kevin Lasagna   
photourl        https://cdn.sofifa.com/players/224/933/21_60.png   
playerurl      http://sofifa.com/player/224933/kevin-lasagna/...   
nationality                                                Italy   
club                                             \n\n\n\nUdinese   
contract                                             2017 ~ 2023   
positions                                                     ST   
height                                                     186cm   
best_position                                                 ST   
loan_date_end                                                NaN   
a/w                                                         High   
d/w                                                       Medium   

                                                           14695  \
name                                                     A. Awad   
longname                                              Ahmed Awad   
photourl        https://cdn.sofifa.com/players/243/159/21_60.png   
playerurl      http://sofifa.com/player/243159/ahmed-awad/210...   
nationality                                            Palestine   
club                                       \n\n\n\nÖstersunds FK   
contract                                             2020 ~ 2021   
positions                                                     ST   
height                                                     182cm   
best_position                                                 ST   
loan_date_end                                                NaN   
a/w                                                       Medium   
d/w                                                       Medium   

                                                           11738  
name                                                   K. Morris  
longname                                           Kieron Morris  
photourl        https://cdn.sofifa.com/players

In a second round of cleaning, I will act on variables: `a/w` and `d/w`

### 9. Set `a/w` and `d/w` to categorical and sort their category values

In [93]:
df['a/w'].isna().sum()

0

In [94]:
df['d/w'].isna().sum()

0

In [95]:
df['a/w'] = df['a/w'].astype('category')
df['d/w'] = df['d/w'].astype('category')

In [96]:
df['a/w'].unique()

['Medium', 'High', 'Low']
Categories (3, object): ['High', 'Low', 'Medium']

In [97]:
df['d/w'].unique()

['Low', 'Medium', 'High']
Categories (3, object): ['High', 'Low', 'Medium']

In [98]:
df['a/w'] = df['a/w'].cat.set_categories(['Low', 'Medium', 'High'], ordered=True)
df['d/w'] = df['d/w'].cat.set_categories(['Low', 'Medium', 'High'], ordered=True)

In [99]:
df['a/w'].sort_values()

8606      Low
3451      Low
7347      Low
12569     Low
15007     Low
         ... 
3767     High
3768     High
14656    High
3758     High
8604     High
Name: a/w, Length: 18979, dtype: category
Categories (3, object): ['Low' < 'Medium' < 'High']

### 10. Clean `contract` variable

In [100]:
df.contract

0        2004 ~ 2021
1        2018 ~ 2022
2        2014 ~ 2023
3        2015 ~ 2023
4        2017 ~ 2022
            ...     
18974    2018 ~ 2022
18975    2020 ~ 2021
18976    2019 ~ 2020
18977    2020 ~ 2022
18978    2019 ~ 2023
Name: contract, Length: 18979, dtype: object

Let's see if there are any other "strange" values that don't start with a year

In [101]:
df.contract[ ~ df.contract.str.startswith('2')]

205      Jun 30, 2021 On Loan
248      Jun 30, 2021 On Loan
254      Jun 30, 2021 On Loan
289                      Free
292                      Free
                 ...         
18472    Aug 31, 2021 On Loan
18571    Jun 30, 2021 On Loan
18600    Dec 31, 2020 On Loan
18622    Dec 31, 2020 On Loan
18680    Dec 31, 2020 On Loan
Name: contract, Length: 1251, dtype: object

I will create a new column called `on_loan_from` for those players who are currently on loan.
For the rest of them, I will create two new columns: `contract_started` and `contract_ends`.  
`contract` column will be reasigned as a boolean column, meaning `free` has a value of False. Otherwise, `True` and will have his/her corresponding `contract_started` and `contract_ends` values.

Now I will create a custom function to deal with different types of contract values.

In [102]:
contract = df.contract.copy()

In [103]:
def process_contract(value):
    # contract has a start and end date
    if '~' in value:
        dates = value.split(' ~ ')
        contract_start = dates[0]
        contract_end = dates[1]
        return [contract_start, contract_end]
    # player has currently no contract
    elif value == 'Free':
        return ['free_player']
    # on loan contract
    else:
        loan_start = value[:-8]
        return ['on_loan', loan_start]      

In [104]:
df.contract.iloc[18472][:-8]

'Aug 31, 2021'

In [105]:
contract = contract.apply(process_contract)

After having applied the function, there will be three types of values:
1. a list with start and end years for contract
2. a list with only one element: 'free'
3. a list with first element equals to 'on_loan' and 2nd element: on_loan_from

Now let's create the three new columns: `contract_started`, `contract_ends`, `has_contract` (boolean) and `on_loan_from`

Create `on_loan` column:

In [106]:
df['on_loan_from'] = [i[1] if i[0] == 'on_loan' else np.nan for i in contract]

In [107]:
df['contract_started'] = [i[0] if len(i[0]) == 4 else np.nan for i in contract]

In [108]:
df['contract_ends'] = [i[1] if len(i[0]) == 4 else np.nan for i in contract]

In [109]:
df['has_contract'] = [bool(i) for i in df.contract_started]

In [110]:
df[['contract_started', 'contract_ends', 'has_contract', 'on_loan_from']]

contract_started contract_ends  has_contract on_loan_from
0                 2004          2021          True          NaN
1                 2018          2022          True          NaN
2                 2014          2023          True          NaN
3                 2015          2023          True          NaN
4                 2017          2022          True          NaN
...                ...           ...           ...          ...
18974             2018          2022          True          NaN
18975             2020          2021          True          NaN
18976             2019          2020          True          NaN
18977             2020          2022          True          NaN
18978             2019          2023          True          NaN

[18979 rows x 4 columns]

The last step is to convert to datetime these new columns

In [111]:
df.contract_started = pd.to_datetime(df.contract_started)
df.contract_ends = pd.to_datetime(df.contract_ends)
df.on_loan_from = df.on_loan_from = pd.to_datetime(df.on_loan_from)

### 11. Clean `club` variable

In [112]:
clubs = df.club.copy()
clubs

0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: club, Length: 18979, dtype: object

In [113]:
clubs = clubs.str.strip()
clubs.sample(5)

8616       Club Libertad
7884     Shonan Bellmare
1741      RSC Anderlecht
7027        Swansea City
13413      Independiente
Name: club, dtype: object

In a previous execution of the cell above, I noticed that some club names have a number at the beginning. I will check and then remove the numbers.

In [114]:
clubs.isna().sum()

0

In [115]:
clubs[clubs.apply(lambda x: x[0].isnumeric() )]

355         1. FC Union Berlin
818         1. FC Union Berlin
967                 1. FC Köln
993                 1. FC Köln
1164                1. FC Köln
                 ...          
17576     1. FC Kaiserslautern
17874        1. FC Saarbrücken
18057    1. FC Heidenheim 1846
18134        1. FC Saarbrücken
18817     1. FC Kaiserslautern
Name: club, Length: 234, dtype: object

234 club names start with '1. '. I Will check if ALL OF THEM start with '1. '

In [116]:
clubs[clubs.str.startswith('1. ')].count()

234

Let's remove the string '1. ' from all of them.

In [117]:
clubs = clubs.str.lstrip('1. ')

In [118]:
clubs[clubs.str.startswith('1. ')].count()

0

All of them have the '1. ' string cleaned!

In [119]:
clubs.sample(20)

10437                   Aragua FC
5938                     SD Aucas
664                       Everton
13401                Warta Poznań
1471               Crystal Palace
11803      Hebei China Fortune FC
4906                Internacional
5781           Club Plaza Colonia
1916             Yeni Malatyaspor
2829                     Elche CF
1444                      No Club
7247                      Farense
18384              Ross County FC
4125     FC Girondins de Bordeaux
16770            Sparta Rotterdam
8993                 ADO Den Haag
3624                 ADO Den Haag
14880       FC Würzburger Kickers
2142             AS Saint-Étienne
14489       AC Mineros de Guayana
Name: club, dtype: object

From a previous execution where  I took a sample of 20 different club values, a value of `No Club` arose. I will replace them by NaN values.

In [120]:
clubs[clubs == 'No Club'].count()

237

In [121]:
clubs[clubs == 'No Club'] = np.nan

Finally, set the cleaned club values to `club` variable in the dataset.

In [122]:
df.club = clubs

### 12. Reestructuration and/or removal of columns after all processing

In [123]:
df.columns

Index(['id', 'name', 'longname', 'photourl', 'playerurl', 'nationality', 'age',
       '↓ova', 'pot', 'club', 'contract', 'positions', 'height', 'weight',
       'preferred_foot', 'bov', 'best_position', 'joined', 'loan_date_end',
       'value_euros_M', 'wage_euros_M', 'release_clause_euros_M', 'attacking',
       'crossing', 'finishing', 'heading_accuracy', 'short_passing', 'volleys',
       'skill', 'dribbling', 'curve', 'fk_accuracy', 'long_passing',
       'ball_control', 'movement', 'acceleration', 'sprint_speed', 'agility',
       'reactions', 'balance', 'power', 'shot_power', 'jumping', 'stamina',
       'strength', 'long_shots', 'mentality', 'aggression', 'interceptions',
       'positioning', 'vision', 'penalties', 'composure', 'defending',
       'marking', 'standing_tackle', 'sliding_tackle', 'goalkeeping',
       'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning',
       'gk_reflexes', 'total_stats', 'base_stats', 'w/f', 'sm', 'a/w', 'd/w',
       'ir', 'pac', 'sho

I will remove the following columns: `contract` and `height`.  
In columns `wage_euros_M`, `release_euros_M` and `value_euros_M`, the string 'euros_M' will be replaced by 'million_euros'.  
Column `weight` will be renamed to `weight_kg`

In [124]:
df.drop(['contract', 'height'], axis=1, inplace=True)

In [125]:
df = df.rename(columns = {'wage_euros_M' : 'wage_million_euros',
                          'release_clause_euros_M' : 'release_clause_million_euros',
                          'value_euros_M' : 'value_million_euros',
                         'weight' : 'weight_kg'})

In [126]:
df.columns[df.columns.str.contains('euros')]

Index(['value_million_euros', 'wage_million_euros',
       'release_clause_million_euros'],
      dtype='object')

# Study of the numeric variables

In [127]:
num_vars = df.select_dtypes('number').columns.to_list()
print(num_vars)

['id', 'age', '↓ova', 'pot', 'weight_kg', 'bov', 'value_million_euros', 'wage_million_euros', 'release_clause_million_euros', 'attacking', 'crossing', 'finishing', 'heading_accuracy', 'short_passing', 'volleys', 'skill', 'dribbling', 'curve', 'fk_accuracy', 'long_passing', 'ball_control', 'movement', 'acceleration', 'sprint_speed', 'agility', 'reactions', 'balance', 'power', 'shot_power', 'jumping', 'stamina', 'strength', 'long_shots', 'mentality', 'aggression', 'interceptions', 'positioning', 'vision', 'penalties', 'composure', 'defending', 'marking', 'standing_tackle', 'sliding_tackle', 'goalkeeping', 'gk_diving', 'gk_handling', 'gk_kicking', 'gk_positioning', 'gk_reflexes', 'total_stats', 'base_stats', 'pac', 'sho', 'pas', 'dri', 'def', 'phy', 'hits', 'height_m', 'height_feet']


In [128]:
df[num_vars].dtypes

id               int64
age              int64
↓ova             int64
pot              int64
weight_kg      float64
                ...   
def              int64
phy              int64
hits           float64
height_m       float64
height_feet    float64
Length: 61, dtype: object

## Study of missing values

In [129]:
def nan_percent(series):
    nan_count = series.isna().sum()
    total_count = series.shape[0]
    return round(nan_count / total_count * 100, 2)

In [130]:
missing_values_pctg = df[num_vars].apply(nan_percent, axis=0).sort_values(ascending=False)
missing_values_pctg.name = 'percentage of missing values'
missing_values_pctg

hits           13.67
id              0.00
gk_diving       0.00
mentality       0.00
aggression      0.00
               ...  
reactions       0.00
balance         0.00
power           0.00
shot_power      0.00
height_feet     0.00
Name: percentage of missing values, Length: 61, dtype: float64

Looks like there are no nan values (except for `hits` variable). That doesn't mean that there are no invalid values. It requires a deeper exploration.  
I will get a sample of the numeric variables, in order to make a visual inspection for strange values.

In [131]:
pd.set_option('display.max_rows', 100)

In [132]:
df[num_vars].sample(10).T

9479           2152           1327   \
id                            239589.000000  236636.000000  225161.000000   
age                               22.000000      23.000000      23.000000   
↓ova                              66.000000      74.000000      76.000000   
pot                               74.000000      82.000000      82.000000   
weight_kg                         76.000000      74.000000      79.000000   
bov                               68.000000      74.000000      78.000000   
value_million_euros                1.900000       9.000000      11.000000   
wage_million_euros                 0.004000       0.019000       0.086000   
release_clause_million_euros       2.100000      16.500000       4.244997   
attacking                        316.000000     315.000000     266.000000   
crossing                          67.000000      72.000000      41.000000   
finishing                         67.000000      53.000000      44.000000   
heading_accuracy                  66.000000      72.000000      74.000000   
short_passing                     61.000000      71.000000      71.000000   
volleys                           55.000000      47.000000      36.000000   
skill                            289.000000     312.000000     256.000000   
dribbling                         68.000000      66.000000      56.000000   
curve                             54.000000      60.000000      29.000000   
fk_accuracy                       37.000000      48.000000      31.000000   
long_passing                      58.000000      69.000000      72.000000   
ball_control                      72.000000      69.000000      68.000000   
movement                         320.000000     344.000000     354.000000   
acceleration                      73.000000      69.000000      67.000000   
sprint_speed                      69.000000      70.000000      72.000000   
agility                           69.000000      70.000000      68.000000   
reactions                         58.000000      73.000000      78.000000   
balance                           51.000000      62.000000      69.000000   
power                            291.000000     321.000000     283.000000   
shot_power                        65.000000      59.000000      49.000000   
jumping                           59.000000      73.000000      77.000000   
stamina                           34.000000      74.000000      54.000000   
strength                          78.000000      73.000000      72.000000   
long_shots                        55.000000      42.000000      31.000000   
mentality                        268.000000     312.000000     269.000000   
aggression                        57.000000      63.000000      80.000000   
interceptions                     29.000000      74.000000      79.000000   
positioning                       64.000000      62.000000      32.000000   
vision                            59.000000      61.000000      33.000000   
penalties                         59.000000      52.000000      45.000000   
composure                         61.000000      68.000000      75.000000   
defending                         52.000000     215.000000     229.000000   
marking                           19.000000      72.000000      78.000000   
standing_tackle                   15.000000      72.000000      75.000000   
sliding_tackle                    18.000000      71.000000      76.000000   
goalkeeping                       55.000000      47.000000      57.000000   
gk_diving                          7.000000      11.000000      10.000000   
gk_handling                       13.000000       9.000000      13.000000   
gk_kicking                        14.000000      12.000000      15.000000   
gk_positioning                     7.000000       7.000000      13.000000   
gk_reflexes                       14.000000       8.000000       6.000000   
total_stats                     1591.000000    1866.000000    1714.000000   
base_stats                       3

The NaN values in `hits` variable may mean that certain profile websites have never been visited. I will fill them with 0 visits.

In [133]:
hits.isna().sum()

2595

In [134]:
hits = hits.fillna(0)

In [135]:
hits[hits == 0].count()

2595

All NaN values from `hits` have been successfully filled with 0 visits values.

## Dealing with outliers

There are too may numeric variables, so I will create a function to check if a variable has outliers.  
A value is considered an outliers if it lies outside the limits of +/- 1.5 · IR (Interquartile Range). 

In [136]:
def has_outliers(series):
    Q1, Q3 = series.quantile([.25, .75])
    IR = Q3 - Q1
    below_Q1 = series < (Q1 - 1.5*IR)
    above_Q3 = series > (Q3 + 1.5*IR)
    number_outliers = (below_Q1 | above_Q3).sum()
    if number_outliers:
        return number_outliers

In [137]:
df[num_vars]

id  age  ↓ova  pot  weight_kg  bov  value_million_euros  \
0      158023   33    93   93       72.0   93               103.50   
1       20801   35    92   92       83.0   92                63.00   
2      200389   27    91   93       87.0   91               120.00   
3      192985   29    91   91       70.0   91               129.00   
4      190871   28    91   91       68.0   91               132.00   
...       ...  ...   ...  ...        ...  ...                  ...   
18974  247223   21    47   55       66.0   49                 0.10   
18975  258760   17    47   67       65.0   51                 0.13   
18976  252757   18    47   65       74.0   49                 0.12   
18977  243790   20    47   57       69.0   48                 0.10   
18978  252520   21    47   57       75.0   50                 0.10   

       wage_million_euros  release_clause_million_euros  attacking  crossing  \
0                   0.560                       138.400        429        85   
1                   0.220                        75.900        437        84   
2                   0.125                       159.400         95        13   
3                   0.370                       161.000        407        94   
4                   0.270                       166.500        408        85   
...                   ...                           ...        ...       ...   
18974               0.001                         0.070        145        23   
18975              50.000                         0.165        211        38   
18976              50.000                         0.131        200        30   
18977               0.002                         0.088        215        45   
18978               0.001                         0.079        163        40   

       finishing  heading_accuracy  short_passing  volleys  skill  dribbling  \
0             95                70             91       88    470         96   
1             95                90             82       86    414         88   
2             11                15             43       13    109         12   
3             82                55             94       82    441         88   
4             87                62             87       87    448         95   
...          ...               ...            ...      ...    ...        ...   
18974         26                43             26       27    142         27   
18975         42                40             56       35    219         46   
18976         34                43             54       39    207         43   
18977         52                34             42       42    194         51   
18978         18                40             45       20    171         40   

       curve  fk_accuracy  long_passing  ball_control  movement  acceleration  \
0         93           94            91            96       451            91   
1         81           76            77            92       431            87   
2         13           14            40            30       307            43   
3         85           83            93            92       398            77   
4         88           89            81            95       453            94   
...      ...          ...           ...           ...       ...           ...   
18974     23           21            29            42       294            68   
18975     40           35            50            48       305            63   
18976     39           31            47            47       290            59   
18977     35           31            31            46       254            62   
18978     33           20            40            38       279            60   

       sprint_speed  agility  reactions  balance  power  shot_power  jumping  \
0                80       91         94       95    389          86       68   
1                91       87         95       71    444          94       95   
2                60       67       

In [138]:
df[num_vars].apply(has_outliers, axis=0).T.sort_values(ascending=False)[:15]

wage_million_euros              4098.0
hits                            2202.0
value_million_euros             2199.0
goalkeeping                     2181.0
gk_kicking                      2089.0
gk_positioning                  2086.0
gk_reflexes                     2085.0
gk_handling                     2085.0
gk_diving                       2085.0
ball_control                    2053.0
release_clause_million_euros    2020.0
attacking                       2016.0
dribbling                       1962.0
short_passing                   1748.0
mentality                       1493.0
dtype: float64

There seems to be a huge range of values for wages, value and hits. Outliers in `hit` column could be due to big differences in popularity (famous players get huge amounts of clicks in their websites, compared to almost none for unknown players). The same for wage and value: wide difference between Messi and a relatively unknown player of a modest football club.

I see many outliers in the goalkeeper variables. Not all players are goalkeepers, so maybe a good idea would be to set NaN at those variables for non-goalkeeper players.

### Soccer / football position abbreviations

Abbreviations for Soccer Midfielder Positions
- CB: Center-back; Stop other team from getting near GK
- GK: Goal Keeper; Stops ball from going in goal
- LB: Left Full-Back; Block other team and hold a line
- LWB: Left Wing-Back; Play near sidelines and block other team
- RB: Right Full-Back; Block other team and hold a line
- RWB: Right Wing-Back; Play near sidelines and block other team
- SW: Sweeper; Stands between GK and everyone else as last line of defense

Midfielders need to be good at both offense and defense because they typically play toward the middle of the field and near their opponent’s goal.

- AM: Attacking Midfielder; Focuses on shooting and scoring
- CM: Center Midfielder; Links defense and offense of a team
- DM: Defensive Midfielder; Organize attacking plays
- LW: Left Winger; Plays along sideline to shoot, assist, and score
- RW: Right Winger; Plays along sideline to shoot, assist, and score

Abbreviations for Soccer Forward Positions

Forwards are the players who play closer to their opponent’s goal and try to score goals.

- CF: Center Forward; Link offense and defense of a team
- S: Striker; Focuses on scoring

In [139]:
goalkeeper_attributes = [i for i in df.columns if i.startswith('gk') ]
df[['positions'] + goalkeeper_attributes]

positions  gk_diving  gk_handling  gk_kicking  gk_positioning  \
0      RW, ST, CF          6           11          15              14   
1          ST, LW          7           11          15              14   
2              GK         87           92          78              90   
3         CAM, CM         15           13           5              10   
4         LW, CAM          9            9          15              15   
...           ...        ...          ...         ...             ...   
18974          CB          7            8           5              14   
18975          CM         12           10           9               6   
18976          CM         13           12           6               6   
18977          RW         14           12           9               8   
18978      CB, LB         10           11           7              10   

       gk_reflexes  
0                8  
1               11  
2               90  
3               13  
4               11  
...            ...  
18974           11  
18975            8  
18976           11  
18977           12  
18978           12  

[18979 rows x 6 columns]

In [140]:
df[ goalkeeper_attributes][df.positions.str.contains('GK')].mean()

gk_diving         65.245783
gk_handling       62.944578
gk_kicking        61.772048
gk_positioning    63.296867
gk_reflexes       66.186506
dtype: float64

Watching the mean values of goalkeeper's attributes, one can see that the presence of outliers in those variables is due to most players not being a goalkeeper. The average values for goalkeepers are higher than those for non-goalkeeper players.

One last step about analyzing outliers: the percentage of outliers. I will take into acount only those variables that have a significative % (% >= 5) of outliers. This minimum threshold can be easily changed in the code below.

In [141]:
outliers_pctg = ( df[num_vars].apply(has_outliers, axis=0) / df.shape[0] * 100 ).dropna()
outliers_pctg = round( outliers_pctg, 2).sort_values(ascending=False)
outliers_pctg[outliers_pctg >= 5]

wage_million_euros              21.59
hits                            11.60
value_million_euros             11.59
goalkeeping                     11.49
gk_kicking                      11.01
gk_reflexes                     10.99
gk_positioning                  10.99
gk_handling                     10.99
gk_diving                       10.99
ball_control                    10.82
release_clause_million_euros    10.64
attacking                       10.62
dribbling                       10.34
short_passing                    9.21
mentality                        7.87
skill                            6.50
heading_accuracy                 5.84
dtype: float64

Watching the results, most of the variable with more or less high percentages are due to the before mentioned reasons, so I won't take action regarding to outliers.

And finally I will export the clean dataframe:

In [142]:
df.to_csv('clean_fifa21_dataset.csv', index=False, na_rep='NULL')